polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -176,32 +207,79 @@ pub enum ArtifactState {
 
 /// A container of all known artifact ids and their states.
 pub struct Artifacts {
-	artifacts: HashMap<ArtifactId, ArtifactState>,
+	inner: HashMap<ArtifactId, ArtifactState>,
 }
 
 impl Artifacts {
-	/// Initialize a blank cache at the given path. This will clear everything present at the
-	/// given path, to be populated over time.
-	///
-	/// The recognized artifacts will be filled in the table and unrecognized will be removed.
-	pub async fn new(cache_path: &Path) -> Self {
-		// First delete the entire cache. This includes artifacts and any leftover worker dirs (see
-		// [`WorkerDir`]). Nodes are long-running so this should populate shortly.
-		let _ = tokio::fs::remove_dir_all(cache_path).await;
+	#[cfg(test)]
+	pub(crate) fn new() -> Self {
+		Self { inner: HashMap::new() }
+	}
+
+	#[cfg(test)]
+	pub(crate) fn len(&self) -> usize {
+		self.inner.len()
+	}
+
+	/// Create a table with valid artifacts and prune the invalid ones.
+	pub async fn new_and_prune(cache_path: &Path) -> Self {
+		let mut artifacts = Self { inner: HashMap::new() };
+		Self::prune_and_insert(cache_path, &mut artifacts).await;
+		artifacts
+	}
+
+	// FIXME eagr: extremely janky, please comment on the appropriate way of setting
+	// * `last_time_needed` set as roughly around startup time
+	// * `prepare_stats` set as Nones, since the metadata was lost
+	async fn prune_and_insert(cache_path: impl AsRef<Path>, artifacts: &mut Artifacts) {
+		fn is_stale(file_name: &str) -> bool {
+			!file_name.starts_with(ARTIFACT_PREFIX)
+		}
+
+		fn insert_cache(artifacts: &mut Artifacts, id: ArtifactId) {
+			let last_time_needed = SystemTime::now();
+			let prepare_stats = Default::default();
+			always!(artifacts
+				.inner
+				.insert(id, ArtifactState::Prepared { last_time_needed, prepare_stats })
+				.is_none());
+		}
+
 		// Make sure that the cache path directory and all its parents are created.
+		let cache_path = cache_path.as_ref();
 		let _ = tokio::fs::create_dir_all(cache_path).await;
 
-		Self { artifacts: HashMap::new() }
-	}
+		if let Ok(mut dir) = tokio::fs::read_dir(cache_path).await {
+			let mut prunes = vec![];
+
+			loop {
+				match dir.next_entry().await {
+					Ok(None) => break,
+					Ok(Some(entry)) => {
+						let file_name = entry.file_name();
+						if let Some(file_name) = file_name.to_str() {
+							if is_stale(file_name) {
+								prunes.push(tokio::fs::remove_file(cache_path.join(file_name)));
+							} else if let Some(id) = ArtifactId::from_file_name(file_name) {
+								insert_cache(artifacts, id);
+							}
```

Comment:
We should have some logs here for both cases:

1. "stale artifact found for old version, discarding"
2. "valid artifact found on-disk for current version, re-using"

For case (1) it would be really nice to log why the artifact was invalid, e.g. log the stale artifact's version.